README.md
# folded/routing

Routing functions for your web application.

[![Packagist License](https://img.shields.io/packagist/l/folded/routing)](https://github.com/folded-php/routing/blob/master/LICENSE) [![Packagist PHP Version Support](https://img.shields.io/packagist/php-v/folded/routing)](https://github.com/folded-php/routing/blob/master/composer.json#L14) [![Packagist Version](https://img.shields.io/packagist/v/folded/routing)](https://packagist.org/packages/folded/routing) [![Build Status](https://travis-ci.com/folded-php/routing.svg?branch=master)](https://travis-ci.com/folded-php/routing) [![Maintainability](https://api.codeclimate.com/v1/badges/0be9fba0b7990aba814d/maintainability)](https://codeclimate.com/github/folded-php/routing/maintainability) [![TODOs](https://img.shields.io/endpoint?url=https://api.tickgit.com/badge?repo=github.com/folded-php/routing)](https://www.tickgit.com/browse?repo=github.com/folded-php/routing)

## Summary

- [About](#about)
- [Features](#features)
- [Requirements](#requirements)
- [Installation](#installation)
- [Examples](#examples)
- [Version support](#version-support)

## About

I created this library to have a standalone, fully featured router, and be able to pull it in any project, that have some existing code or not.

Folded is a constellation of packages to help you setting up a web app easily, using ready to plug in packages.

- [folded/action](https://github.com/folded-php/action): A way to organize your controllers for your web app.
- [folded/config](https://github.com/folded-php/config): Configuration utilities for your PHP web app.
- [folded/crypt](https://github.com/folded-php/crypt): Encrypt and decrypt strings for your web app.
- [folded/exception](https://github.com/folded-php/exception): Various kind of exception to throw for your web app.
- [folded/history](https://github.com/folded-php/history): Manipulate the browser history for your web app.
- [folded/http](https://github.com/folded-php/http): HTTP utilities for your web app.
- [folded/orm](https://github.com/folded-php/orm): An ORM for you web app.
- [folded/request](https://github.com/folded-php/request): Request utilities, including a request validator, for your PHP web app.
- [folded/session](https://github.com/folded-php/session): Session functions for your web app.
- [folded/view](https://github.com/folded-php/view): View utilities for your PHP web app.

## Features

- register GET and POST routes actions (using callback)
- can match the current browsed URL and execute the associated callback
- can name the registered route to use additional functions like:
  - getting the URL from a route name

## Requirements

- PHP version 7.4+
- Composer installed

## Installation

- [1. Install the package](1-install-the-package)
- [2. Bootstrap the router](#2-bootstrap-the-router)

### 1. Install the package

On your project root directory, run this command:

```bash
composer require folded/routing
```

### 2. Bootstrap the router

In the script that is called first, register your routes, and then call for `matchRequestedUrl()`:

```php
use function Folded\addGetRoute;
use function Folded\matchRequestedUrl;

addGetRoute("/", function() {
  echo "Hello world";
});

try {
    matchRequestedUrl();
} catch (Exception $exception) {
  // ...
}
```

## Examples

As this library is using [nikic/fast-route](https://github.com/nikic/FastRoute) internally, refer to this documentation to know [all the possibilities](https://github.com/nikic/FastRoute#defining-routes) regarding constructing the route string.

- [1. Register a GET route](#1-register-a-get-route)
- [2. Register a POST route](#2-register-a-post-route)
- [3. Catching url not found exceptions](#3-catching-url-not-found-exceptions)
- [4. Catching method not allowed exceptions](#4-catching-method-not-allowed-exceptions)
- [5. Naming a route](#5-naming-a-route)
- [6. Get the URL of a named route](#6-get-the-url-of-a-named-route)
- [7. Redirect to a named route](#7-redirect-to-a-named-route)
- [8. Redirect to an URL](#8-redirect-to-an-url)
- [9. Check if the current URL is the given one](#9-check-if-the-current-url-is-the-given-one)
- [10. Check if a route matches the current URL](#10-check-if-a-route-matches-the-current-url)

### 1. Register a GET route

In this example, we will register a GET route.

```php
use function Folded\addGetRoute;

addGetRoute("/about-us", function() {
  echo "<h1>About us</h1>";
});
```

### 2. Register a POST route

In this example, we will register a POST route.

```php
use function Folded\addPostRoute;

addPostRoute("/search/{search}", function($search) {
  // Pulling posts from the database...
  echo "<h1>Search result for $search</h1>";
});
```

### 3. Catching url not found exceptions

In this example, we will catch a not found exception.

```php
use function Folded\matchRequestedUrl;
use Folded\Exceptions\UrlNotFoundException;

try {
  matchRequestedUrl();
} catch (Exception $exception) {
  if ($exception instanceof UrlNotFoundException) {
    // Log it, or send it to an error management system...
    // Display a 404 page...
  }
}
```

### 4. Catching method not allowed exceptions

In this example, we will catch a method not allowed error (which happens when you browsed an URL, but this url has been registered with another protocol).

```php
use function Folded\matchRequestedUrl;
use Folded\Exceptions\MethodNotAllowedException;

try {
  matchRequestedUrl();
} catch (Exception $exception) {
  if ($exception instanceof MethodNotAllowedException) {
    // Log it, or send it to an error management system...
    // Display a 405 page...
  }
}
```

### 5. Naming a route

In this example, we will name a route.

```php
use function Folded\addGetRoute;

addGetRoute("/", function() {
  echo "welcome home";
}, "home.index");
```

### 6. Get the URL of a named route

In this example, we will get the name of a named route.

```php
use function Folded\addGetRoute;
use function Folded\getRouteUrl;

addGetRoute("/user/{user}/post/{post}", function($user, $post) {
  echo "User $user posted $post";
}, "user.post.show");

echo getRouteUrl("user.post.show", ["user" => 42, "post" => 1]); // string(15) "/user/42/post/1"
```

### 7. Redirect to a named route

In this example, we will redirect to the URL of a named route.

```php
use function Folded\addGetRoute;
use function Folded\redirectToRoute;

addGetRoute("/about-us", function() {
  echo "<h1>About us</h1>";
}, "about-us.index");

redirectToRoute("about-us.index");
```

By default, a status code `303` will be used alongside the redirection. You can override this behavior by adding the HTTP status code of your choice as second parameter:

```php
use function Folded\addGetRoute;
use function Folded\redirectToRoute;

addGetRoute("/about-us", function() {
  echo "<h1>About us</h1>";
}, "about-us.index");

redirectToRoute("about-us.index", 200);
```

### 8. Redirect to an URL

In this example, we will redirect to a plain URL.

```php
use function Folded\redirectToUrl;

redirectToUrl("/about-us");
```

By default, a status code `303` will be used alongside the redirection. You can override this behavior by adding the HTTP status code of your choice as second parameter:

```php
use function Folded\redirectToUrl;

redirectToUrl("/", 200);
```

### 9. Check if the current URL is the given one

In this example, we will check if the current url is the one we have in parameter.

```php
use function Folded\currentUrlIs;

if (currentUrlIs("/")) {
  echo "we are at the home page";
} else {
  echo "we are not at the home page";
}
```

### 10. Check if a route matches the current URL

In this example, we will check if a given route name matches the URL.

```php
use function Folded\currentRouteIs;
use function Folded\addGetRoute;

addGetRoute("/about-us", function() {
  echo "<h1>About us</h1>";
}, "about-us.index");

if (currentRouteIs("about-us.index")) {
  echo "this is the about us page";
} else {
  echo "this is not the about us page";
}
```

Note that if your route contains named parameter, for example `/user/{user}`, you can use the second parameter to fill the named parameter with the actual value.

```php
use function Folded\currentRouteIs;
use function Folded\addGetRoute;

addGetRoute("/user/{user}", function() {
  echo "<h1>User detail</h1>";
}, "user.show");

if (currentRouteIs("user.show", ["user" => 17])) {
  echo "this is the route /user/17";
} else {
  echo "this is not the route /user/17";
}
```

## Version support

|        | 7.3 | 7.4 | 8.0 |
| ------ | --- | --- | --- |
| v0.1.0 | ❌  | ✔️  | ❓  |
| v0.1.1 | ❌  | ✔️  | ❓  |
| v0.1.2 | ❌  | ✔️  | ❓  |
| v0.1.3 | ❌  | ✔️  | ❓  |
| v0.2.0 | ❌  | ✔️  | ❓  |
| v0.3.0 | ❌  | ✔️  | ❓  |
| v0.4.0 | ❌  | ✔️  | ❓  |
| v0.4.1 | ❌  | ✔️  | ❓  |
| v0.5.0 | ❌  | ✔️  | ❓  |
| v0.5.1 | ❌  | ✔️  | ❓  |
| v0.6.0 | ❌  | ✔️  | ❓  |


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.6.0] 2020-10-03

### Added

- New method `MethodNotAllowed::getUrl()` to be able to catch the URL when a 405 exception is catched when using `matchRequestedUrl()`.

### Fixed

- Calling `currentRouteIs()` will now return true even if the current URL has query strings.\*
- Updated dependency `folded/exception` to version 0.4.\*.

## [0.5.1] 2020-10-01

### Fixed

- Added missing namespace `Folded` in `function_exists` statements.

## [0.5.0] 2020-09-23

### Added

- New method `currentRouteIs` to check if a given route matches the current URL.
- New method `currentUrlIs` to check if a given URL matches the current URL.

## [0.4.1] 2020-09-18

### Fixed

- Dependency updated.

## [0.4.0] 2020-09-18

### Breaking

- The function `redirectToRoute()` will now throw a `Folded\Exceptions\RouteNotFoundException` instead of a `OutOfRangeException` in case the route is not found.

## [0.3.0] 2020-09-15

### Added

- New method `redirectToRoute("home.index")` which redirect to the URL matching the route name.
- New method `redirectToUrl("/about-us")` which redirects to a plain URL.

## [0.2.0] 2020-09-12

### Added

- New third optional parameter `$name` for the methods `addGetRoute()` and `addPostRoute()` to name a route.
- New method `getRouteUrl("home.index")` to get the URL from a route name.

## [0.1.3] 2020-09-07

### Added

- **Nothing change** on your code. We moved from local exception to shared exception internally for both `Folded\Exceptions\UrlNotFoundException` and `Folded\Exception\MethodNotAllowedException`.

## [0.1.2] 2020-09-05

### Fixed

- Bug when `addGetRoute()` and `addPostRoute()` were not found.

## [0.1.1] 2020-09-05

### Fixed

- On an IDE that auto import namespaces, exceptions will now be imported prefixed with Folded namespace (instead of another one).

## [0.1.0] 2020-09-05

### Added

- First working version.


src/Router.php
<?php

declare(strict_types = 1);

namespace Folded;

use Closure;
use Http\Redirection;
use OutOfRangeException;
use FastRoute\Dispatcher;
use FastRoute\RouteCollector;
use InvalidArgumentException;
use FastRoute\RouteParser\Std;
use function Foled\getRequestedUri;
use function FastRoute\simpleDispatcher;
use Folded\Exceptions\UrlNotFoundException;
use Folded\Exceptions\MethodNotAllowedException;
use Folded\Exceptions\RouteNotFoundException;

/**
 * Represent the engine that matches the current browsed URL against registered URLs.
 *
 * @since 0.1.0
 */
class Router
{
    /**
     * @since 0.1.0
     */
    const ALLOWED_PROTOCOLS = [
        self::PROTOCOL_GET,
        self::PROTOCOL_POST,
    ];

    /**
     * @since 0.1.0
     */
    const PROTOCOL_GET = "GET";

    /**
     * @since 0.1.0
     */
    const PROTOCOL_POST = "POST";

    /**
     * Stores the engine that is responsible for the url matching logic.
     *
     * @var Dispatcher
     */
    private static ?Dispatcher $engine = null;

    /**
     * Stores the registered routes.
     *
     * @since 0.1.0
     */
    private static array $routes = [];

    /**
     * Add a GET route.
     *
     * @param string  $route  The URL or route pattern.
     * @param Closure $action The function to trigger when the current browsed URL matches the route.
     * @param string  $name   The name of the route (optional).
     *
     * @throws InvalidArgumentException If the route name is empty.
     *
     * @since 0.1.0
     *
     * @example
     * Router::addGetRoute("/", function() {
     *  echo "<h1>Welcome</h1>";
     * });
     */
    public static function addGetRoute(string $route, Closure $action, ?string $name = null): void
    {
        static::addRoute(self::PROTOCOL_GET, $route, $action, $name);
    }

    /**
     * Add a POST route.
     *
     * @param string  $route  The URL or route pattern.
     * @param Closure $action The function to trigger when the current browsed URL matches.
     * @param string  $name   The name of the route (optional).
     *
     * @throws InvalidArgumentException If the route name is empty.
     *
     * @since 0.1.0
     *
     * @example
     * Router::addPostRoute("/search/{search}", function($search) {
     *  echo "search results for $search";
     * });
     */
    public static function addPostRoute(string $route, Closure $action, ?string $name = null): void
    {
        static::addRoute(self::PROTOCOL_POST, $route, $action, $name);
    }

    /**
     * Clears the object, like it has never been called before.
     * Useful for unit testing.
     *
     * @since 0.1.0
     *
     * @example
     * Router::clear();
     */
    public static function clear(): void
    {
        static::$routes = [];
        static::$engine = null;
    }

    /**
     * Returns true if route matches the current URL, else returns false.
     *
     * @param string $route      The route name to verify.
     * @param array  $parameters The parameters to pass to the route if it has named parameters.
     *
     * @throws Folded\Exceptions\RouteNotFoundException If the route name is not found.
     *
     * @since 0.5.0
     *
     * @example
     * if (Router::currentRouteIs("home.index")) {
     *  echo "current route is /";
     * } else {
     *  echo "current route is not /";
     * }
     */
    public static function currentRouteIs(string $route, array $parameters = []): bool
    {
        if (!self::hasRoute($route)) {
            throw (new RouteNotFoundException("route $route not found"))->setRoute($route);
        }

        return self::getRouteUrl($route, $parameters) === parse_url($_SERVER["REQUEST_URI"], PHP_URL_PATH);
    }

    /**
     * Returns true if the URL matches the current URL, else returns false.
     *
     * @param string $url The URL to verify.
     *
     * @throws InvalidArgumentException If the URL is empty.
     *
     * @since 0.5.0
     *
     * @example
     * if (Router::currentUrlIs("/")) {
     *  echo "current URL is /";
     * } else {
     *  echo "current URL is not "/";
     * }
     */
    public static function currentUrlIs(string $url): bool
    {
        if (empty(trim($url))) {
            throw new InvalidArgumentException("url must not be empty");
        }

        return $url === $_SERVER["REQUEST_URI"];
    }

    /**
     * Get the list of registered routes.
     *
     * @since 0.1.0
     *
     * @example
     * Router::getRoutes();
     */
    public static function getRoutes(): array
    {
        return static::$routes;
    }

    /**
     * Get the URL associated with the given route.
     * If the route contains placeholders, you need to pass the parameters to be filled.
     *
     * @param string $name       The name of the route.
     * @param array  $parameters The parameters, by key-value pairs or simply values, to fill in the placeholders (optional).
     *
     * @throws OurOfRangeException If the route name is not found.
     *
     * @since 0.2.0
     *
     * @example
     * Router::getRouteUrl("user.show", ["user" => 42]);
     */
    public static function getRouteUrl(string $name, array $parameters = []): string
    {
        // Use Folded\RouteNotFoundException instead.
        if (!self::hasRoute($name)) {
            throw new OutOfRangeException("route $name not found");
        }

        /**
         * @var Route
         */
        $route = self::$routes[$name];

        return self::replaceRouteParameters($route->getRoute(), $parameters);
    }

    /**
     * Register route and try to match the current browsed URL.
     *
     * @since 0.1.0
     *
     * @throws Folded\Exceptions\UrlNotFoundException      If the URL is not found in the registered routes.
     * @throws Folded\Exceptions\MethodNotAllowedException If the URL has been found, but the current protocol does not match.
     *
     * @example
     * Router::matchRequestedUrl();
     */
    public static function matchRequestedUrl()
    {
        static::startEngine();

        return static::matchUrl();
    }

    /**
     * Redirects to the URL found by its route name.
     *
     * @param string $name   The name of the route.
     * @param int    $status The HTTP status code to use when redirecting (default: 303 - See other).
     *
     * @throws RouteNotFoundException If the route name is not found.
     *
     * @since 0.3.0
     *
     * @example
     * Router::addGetRoute("/", function() {}, "home.index");
     *
     * Router::redirectToRoute("home.index"); // redirects to "/"
     */
    public static function redirectToRoute(string $name, int $status = Redirection::SEE_OTHER): void
    {
        if (!self::hasRoute($name)) {
            throw (new RouteNotFoundException("route $name not found"))->setRoute($name);
        }

        $url = self::getRouteUrl($name);

        http_response_code($status);

        header("Location:$url");
    }

    /**
     * Redirects to a given URL.
     *
     * @param string $url    The URL to redirect to.
     * @param int    $status The HTTP status code ot use when redirecting.
     *
     * @since 0.3.0
     *
     * @example
     * Router::redirectToUrl("/about-us");
     */
    public static function redirectToUrl(string $url, int $status = Redirection::SEE_OTHER): void
    {
        http_response_code($status);

        header("Location:$url");
    }

    /**
     * Registers route into the engine.
     *
     * @since 0.1.0
     *
     * @example
     * Router::startEngine();
     */
    public static function startEngine(): void
    {
        self::$engine = simpleDispatcher(static function (RouteCollector $router): void {
            foreach (static::$routes as $route) {
                $router->addRoute($route->getProtocol(), $route->getRoute(), $route->getAction());
            }
        });
    }

    /**
     * Add a route to the list of registered routes.
     *
     * @param string  $protocol The HTTP protocol.
     * @param string  $route    The URL or route pattern.
     * @param Closure $action   The function to trigger when the current browsed URL matches the route.
     * @param string  $name     The name of the route (optional).
     *
     * @throws InvalidArgumentException If the route name is empty.
     *
     * @since 0.1.0
     *
     * @example
     * Router::addRoute("GET", "/", fn () => echo "hello world");
     */
    private static function addRoute(string $protocol, string $route, Closure $action, ?string $name = null): void
    {
        $route = new Route($protocol, $route, $action);

        if ($name !== null) {
            if (empty(trim($name))) {
                throw new InvalidArgumentException("route name cannot be empty");
            }

            static::$routes[$name] = $route;
        } else {
            static::$routes[] = $route;
        }
    }

    /**
     * Returns true if the route name is found, else returns false.
     *
     * @param string $name The route name.
     *
     * @since 0.3.0
     *
     * @example
     * if (Router::hasRoute("home.index")) {
     *  echo "has route home.index";
     * } else {
     *  echo "has not route home.index";
     * }
     */
    private static function hasRoute(string $name): bool
    {
        return isset(self::$routes[$name]);
    }

    /**
     * Make the engine to match the current browsed URL.
     *
     * @since 0.1.0
     *
     * @throws Folded\Exceptions\UrlNotFoundException      If the current browsed URL does not match the registered URLs.
     * @throws Folded\Exceptions\MethodNotAllowedException If the current browsed URL matches, but the protocol don't.
     *
     * @example
     * Router::matchUrl();
     */
    private static function matchUrl()
    {
        $requestMethod = getRequestedMethod();
        $uri = getRequestedUri();

        $routeInfo = static::$engine->dispatch($requestMethod, $uri);

        switch ($routeInfo[0]) {
            case Dispatcher::NOT_FOUND:
                throw (new UrlNotFoundException("url not found"))->setUrl($uri);
            case Dispatcher::METHOD_NOT_ALLOWED:
                $allowedMethods = $routeInfo[1];

                throw (new MethodNotAllowedException("method not allowed"))->setMethodNotAllowed($requestMethod)->setAllowedMethods($allowedMethods)->setUrl($uri);
            case Dispatcher::FOUND:
                $handler = $routeInfo[1];
                $vars = $routeInfo[2];

                return call_user_func_array($handler, $vars);
        }
    }

    /**
     * Replace the placeholders in a route by the given parameters (as either key-value pairs or a list of values).
     *
     * @param string $route      The route to replace parameter on. It is NOT the route name.
     * @param array  $parameters A key-value pairs or a list of value to use to replace the placeholders.
     *
     * @throws InvalidArgumentException If a parameter to replace is missing for the route.
     * @throws InvalidArgumentException If a parameter does not match the expected regex in the route.
     *
     * @since 0.2.0
     *
     * @example
     * echo Router::replaceRouteParameter("/user/{user}/post/{post}", [
     *  "user" => 42,
     *  "post" => 2
     * ]);
     */
    private static function replaceRouteParameters(string $route, array $parameters): string
    {
        $routeDatas = (new Std())->parse($route);
        $placeholders = $parameters;

        $url = "";

        foreach ($routeDatas as $routeData) {
            foreach ($routeData as $data) {
                if (is_string($data)) {
                    // This is a string, so nothing to replace inside of it
                    $url .= $data;
                } elseif (is_array($data)) {
                    // This is an array, so it contains in first the name of the parameter, and in second the regular expression.
                    // Example, [0 => "name", 1 => "[^/]"]
                    [$parameterName, $regularExpression] = $data;

                    $parameterValue = null;

                    if (isset($placeholders[$parameterName])) {
                        // If the parameter name is found by its key in the $parameters parameter, we use it
                        $parameterValue = $placeholders[$parameterName];

                        // We remove it from the remaining placeholders values
                        unset($placeholders[$parameterName]);
                    } elseif (isset($placeholders[0])) {
                        // Else, we take the first parameter in the $parameters parameter
                        $parameterValue = $placeholders[0];

                        // We remove it from the remaining available placeholders values
                        array_shift($placeholders);
                    } else {
                        throw new InvalidArgumentException("parameter $parameterName missing for route $route");
                    }

                    // Checking if the value found matches the regular expression of the associated route parameter
                    $matches = [];
                    $success = preg_match("/" . str_replace("/", "\/", $regularExpression) . "/", (string) $parameterValue, $matches);

                    if ($success !== 1 || (isset($matches[0]) && $parameterValue != $matches[0])) {
                        throw new InvalidArgumentException("parameter $parameterName does not matches regular expression $regularExpression for route $route");
                    }

                    $url .= $parameterValue;
                }
            }
        }

        return $url;
    }
}
